File: tic-tac-toe-javafx-java/src/main/java/com/mj/tic/tac/toe/javafx/java/task/BaseTask.java
package com.mj.tic.tac.toe.javafx.java.task;

import javafx.concurrent.Task;

/**
 * Package-private abstract base class for all application background tasks.
 *
 * <p>This class extends {@link javafx.concurrent.Task Task&lt;V&gt;} and
 * redeclares {@link #call()} as abstract, providing a common supertype for
 * every task executed within the application's mediator-executor pipeline.
 * Concrete subclasses implement specific game operations such as marking a
 * cell, computing an AI move, or handling game-over presentation.</p>
 *
 * <p>Tasks are executed through the
 * {@link com.mj.tic.tac.toe.javafx.java.controller.ControllerMediator
 * ControllerMediator} interface and are submitted to a shared single-thread
 * executor, ensuring sequential, non-overlapping background work.</p>
 *
 * <p><b>Package-private visibility</b> &mdash; only classes within the
 * {@code task} package may extend this base. All application tasks are
 * expected to reside in this package.</p>
 *
 * @param <V> the result type returned by the task's {@link #call()} method
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @see com.mj.tic.tac.toe.javafx.java.task.MarkCellTask
 * @see com.mj.tic.tac.toe.javafx.java.task.ComputerMoveTask
 * @see com.mj.tic.tac.toe.javafx.java.task.GameOverTask
 * @see com.mj.tic.tac.toe.javafx.java.task.PlayGameTask
 * @see com.mj.tic.tac.toe.javafx.java.task.ResetGameTask
 * @since 20-01-2023
 */

abstract class BaseTask<V> extends Task<V> {

    /**
     * Invoked on a background thread to perform the task's work.
     *
     * <p>Subclasses must implement this method to define the actual work
     * performed off the JavaFX Application Thread. The return value is
     * delivered to any {@link javafx.concurrent.Task#setOnSucceeded
     * onSucceeded} handler registered on this task.</p>
     *
     * @return the computed result, or {@code null} if the task produces
     * no value
     * @throws Exception if the background computation fails
     */
    @Override
    protected abstract V call() throws Exception;
}

File: tic-tac-toe-javafx-kotlin/src/main/kotlin/com/mj/tic/tac/toe/javafx/kotlin/task/BaseTask.kt
package com.mj.tic.tac.toe.javafx.kotlin.task

import javafx.concurrent.Task

/**
 * Abstract base class for all background tasks in the application.
 *
 * Extends [javafx.concurrent.Task] to provide a common supertype
 * for all game-related background operations. Subclasses implement
 * [call] to perform their work on a background thread and return
 * a result of type [V].
 *
 * Tasks are executed via the [com.mj.tic.tac.toe.javafx.kotlin.controller.ControllerMediator] (implemented by
 * [com.mj.tic.tac.toe.javafx.kotlin.controller.view.ViewController]) on a single-thread executor, ensuring sequential
 * execution and preventing concurrency issues in the game logic.
 *
 * @param V The result type produced by the task.
 * @author Montaser Sbaih
 * @version 1.0
 * @email montaser.jjs@gmail.com
 * @phone +962-786258874
 * @since 20-01-2023
 */

abstract class BaseTask<V> : Task<V>() {

    /**
     * Performs the background computation and returns the result.
     *
     * Called by the JavaFX concurrency framework on a background
     * thread. Subclasses must implement this method to perform
     * their specific game logic.
     *
     * @return The result of type [V], or null if applicable.
     * @throws Exception if the computation fails.
     */
    @Throws(Exception::class)
    abstract override fun call(): V
}